app/services/title_builder.rb

```diff
@@ -4,38 +4,130 @@ class TitleBuilder
+        structured_title = if structured_title.present?
+                             "#{structured_title.sub(/[. :]+$/, '')} : #{value.sub(/^:/, '').strip}"
+                           else
+                             structured_title = value.sub(/^:/, '').strip
+                           end
+      end
+    end
+    structured_title
+  end
+  # rubocop:enable Metrics/BlockLength
+  # rubocop:enable Metrics/CyclomaticComplexity
+  # rubocop:enable Metrics/MethodLength
+  # rubocop:enable Metrics/PerceivedComplexity
+  private_class_method :title_from_structured_values
+
+  def self.remove_trailing_punctuation(title)
+    title.sub(%r{[ .,;:/\\]+$}, '')
+  end
+  private_class_method :remove_trailing_punctuation
+
+  # @param [Array<Cocina::Models::Title>] titles
+  # @return [Cocina::Models::Title, nil] title that has status=primary
+  def self.primary_title(titles)
+    primary_title = titles.find do |title|
+      title.status == 'primary'
+    end
+    return primary_title if primary_title.present?
+
+    # NOTE: structuredValues would only have status primary assigned as a sibling, not as an attribute
+
+    titles.find do |title|
+      title.parallelValue&.find do |parallel_title|
+        parallel_title.status == 'primary'
+      end
+    end
+  end
+  private_class_method :primary_title
+
+  # @param [Array<Cocina::Models::Title>] titles
+  # @return [Cocina::Models::Title, nil] first title that has no type attribute
+  def self.first_untyped_title(titles)
+    titles.find do |title|
+      if title.parallelValue.present?
+        title.parallelValue&.find { |parallel_value| parallel_value.type.nil? }
       else
-        [title.value, join_token(title.type)]
+        title.type.nil?
       end
     end
-    last = title_parts.pop
-    title_parts.join + last.first # Drops the join token from the last element
   end
+  private_class_method :first_untyped_title
+
+  def self.non_sorting_char_count(title)
+    non_sort_note = title.note&.find { |note| note.type&.downcase == 'nonsorting character count' }
+    return 0 unless non_sort_note
 
-  def self.join_token(type)
-    case type
-    when 'nonsorting characters'
-      ' '
-    when 'subtitle'
-      '. '
-    when 'part number'
-      ', '
-    else
-      ' : '
+    non_sort_note.value.to_i
+  end
+  private_class_method :non_sorting_char_count
+
+  # combine part name and part number:
+  #   respect order of occurrence
+  #   separated from each other by comma space
+  # rubocop:disable Metrics/PerceivedComplexity
+  def self.title_from_structured_part(structured_values)
+    title_from_part = ''
+    structured_values.each do |structured_value|
+      case structured_value.type&.downcase
+      when 'part name', 'part number'
+        title_from_part = if title_from_part&.strip.present?
+                            "#{title_from_part.sub(/[ .,]*$/, '')}, #{structured_value.value&.strip}"
```

Comment:
why is `value&.strip` used here? Are you expecting it to be nil?

Reply:
just in case.  I've found that data is often dirty.

Reply:
this looks like it could result in a "title_part_value, " (with comma blank at the end) if the structured value is nil. Is that ok?

Reply:
any trailing punct is going to be removed with remove_trailing_punctuation as the last line of the public build method